### PR TITLE
Fix targeted shrinking in USERNF

### LINKED ISSUE #221: No Shrink in TPBT
While experimenting with TPBT properties, I came across some cases where shrinking was not happening at all for the generators, when using the `?USERNF` macro.

## Example

```erlang
-module(bad_shrink).
-include_lib("proper/include/proper.hrl").
-compile(export_all).

%% -----------------------------------------------------------------------------
%% Generators
%% -----------------------------------------------------------------------------

normal_int() ->
  integer().

let_int() ->
  ?LET(I, integer(), I).

nf_int() ->
  fun (Prev, _T) ->
    ?LET(I, integer(Prev, inf), I)
  end.

normal_int_user_nf() ->
  ?USERNF(normal_int(), nf_int()).

let_int_user_nf() ->
  ?USERNF(let_int(), nf_int()).

normal_list() ->
  list(integer()).

let_list() ->
  ?LET(L, list(integer()), L).

nf_list() ->
  fun (Prev, _T) ->
    Max = case Prev of
      [] -> 0;
      _ -> lists:max(Prev)
    end,
    ?LET(L, vector(length(Prev), integer(Max, inf)), Prev ++ L)
  end.

normal_list_user_nf() ->
  ?USERNF(normal_list(), nf_list()).

let_list_user_nf() ->
  ?USERNF(let_list(), nf_list()).

%% -----------------------------------------------------------------------------
%% Properties
%% -----------------------------------------------------------------------------

property_int(I) ->
  ?MAXIMIZE(I),
  I < 500.

prop_normal_int() ->
  ?FORALL_TARGETED(I, normal_int_user_nf(), property_int(I)).

prop_let_int() ->
  ?FORALL_TARGETED(I, let_int_user_nf(), property_int(I)).

property_list(L) ->
  ?MAXIMIZE(lists:sum(L)),
  lists:sum(L) < 1000.

prop_normal_list() ->
  ?FORALL_TARGETED(L, normal_list_user_nf(), property_list(L)).

prop_let_list() ->
  ?FORALL_TARGETED(L, let_list_user_nf(), property_list(L)).
```

### Output

```
1> c(bad_shrink).
bad_shrink.erl:3: Warning: export_all flag enabled - all functions will be exported
{ok,bad_shrink}
2> proper:quickcheck(bad_shrink:prop_normal_int()).
.....................!
Failed: After 22 test(s).
513

Shrinking .(1 time(s))
500
false
3> proper:quickcheck(bad_shrink:prop_let_int()).
..............................................!
Failed: After 47 test(s).
559

Shrinking (0 time(s))
559
false
4> proper:quickcheck(bad_shrink:prop_normal_list()).
..!
Failed: After 3 test(s).
[-2,-7,0,-1,3,-9,21,29,8,3,5,6,29,33,29,32,35,33,35,33,37,33,30,29,46,38,37,46,38,50,71,45,49,39,45,38,49,39,40,42,47,44,41,43,42,44,52,39]

Shrinking ...(3 time(s))
[20,37,46,38,50,71,45,49,39,45,38,49,39,40,42,47,44,41,43,42,44,52,39]
false
5> proper:quickcheck(bad_shrink:prop_let_list()).
...!
Failed: After 4 test(s).
[23,8,34,36,44,45,49,48,50,49,55,50,55,52,56,54,61,60,67,60,56,60,57,60,71,75,56,64,62,61,59,59]

Shrinking (0 time(s))
[23,8,34,36,44,45,49,48,50,49,55,50,55,52,56,54,61,60,67,60,56,60,57,60,71,75,56,64,62,61,59,59]
false
```

## Problem

The problem seems to come from the way `?USERNF`s are treated when shrinking. Specifically in [`proper.erl`](https://github.com/proper-testing/proper/blob/a0222a73527ecadcee9472fecc8b1cbaa56547a8/src/proper.erl#L1776) where if the generator is a `user_nf` the instance is cleaned. This works well if the `first` generator is of a simple type (`integer()`, `float()`, etc.), but not so well if both generators `first` and `next` produce same instances (not simple).

## Solution

Lines 1776-1781 in `proper.erl` could become:

```erlang
ImmInstance = case proper_gen:safe_is_instance(RawImmInstance, Type) of
  false -> proper_gen:clean_instance(RawImmInstance);
  _ -> RawImmInstance
end,
```

However, this would not solve opposite cases, where the `first` generator is complex (`?LET` for example) and the `next` function is a simple generator.

### PR DESCRIPTION
**Pull request by xspirus**
This commit fixes #221 and updates the way the targeted implementation
shrinks its instances. PropEr now tries to find the best shrinker
between the original generator and the neighborhood function.

Users can now also specify `?SHRINK` macros in the neighborhood
function, which will take precedence.

### CODE REVIEW

File: include/proper_common.hrl [hunk 1]
```diff
@@ -34,8 +34,7 @@
 -define(FORALL(X,RawType,Prop), proper:forall(RawType, fun(X) -> Prop end)).
 -define(EXISTS(X,RawType,Prop), proper:exists(RawType, fun(X) -> Prop end, false)).
 -define(NOT_EXISTS(X,RawType,Prop), proper:exists(RawType, fun(X) -> Prop end, true)).
--define(FORALL_TARGETED(X, RawType, Prop),
-        proper:targeted(RawType, fun(X) -> Prop end)).
```
Comment (kostis) [inline include/proper_common.hrl]: I think this was written this way because we do not want lines > 80 chars

Comment (xspirus) [reply include/proper_common.hrl]: Yes will update with fixes as well.

File: src/proper_target.erl [hunk 2]
```diff
@@ -265,6 +274,12 @@ handle_call(gen, _From, State) ->
   Data = State#state.data,
   {NextValue, NewTarget, NewData} = Strat:next(Target, Data),
   {reply, NextValue, State#state{target = NewTarget, data = NewData}};
+handle_call(shrinker, _From, State) ->
+  Strat = State#state.strategy,
```
Comment (kostis) [inline src/proper_target.erl]: These three lines that access these three fields can be written in just one.

File: src/proper.erl [hunk 3]
```diff
@@ -1774,12 +1751,14 @@ shrink(Shrunk, [ImmInstance | Rest], {_Type,Prop}, Reason,
...
-		      {ok, _} ->
-			  proper_gen:clean_instance(RawImmInstance);
-		      error ->
-			  RawImmInstance
-		  end,
+    ImmInstance = case get('$user_nf') of
+                    true ->
+                      case proper_types:safe_is_instance(RawImmInstance, Type) of
+                        false -> proper_gen:clean_instance(RawImmInstance);
+                        _ -> RawImmInstance
```
Comment (kostis) [inline src/proper.erl]: this `_' should be `true`, I think, unless you really want it to match with `{error,{typeserver,...}}` that `safe_is_instance` also returns.

File: src/proper.erl [hunk 4]
```diff
@@ -1774,12 +1751,14 @@ shrink(Shrunk, [ImmInstance | Rest], {_Type,Prop}, Reason,
...
+                      end;
+                    _ -> RawImmInstance
```
Comment (kostis) [inline src/proper.erl]: Similarly to the above, the `_` here should read `false` for avoiding this case matching if `$user_nf` is not set.

Comment (xspirus) [reply src/proper.erl]: This could be `undefined` if `$user_nf` is not set at all.

File: src/proper.erl [hunk 5]
```diff
@@ -1774,12 +1741,15 @@ shrink(Shrunk, [ImmInstance | Rest], {_Type,Prop}, Reason,
...
-		      error ->
-			  RawImmInstance
-		  end,
+    ImmInstance = case proper_types:find_prop(is_user_nf, Type) of
+                    {ok, true} ->
+                      case proper_types:safe_is_instance(RawImmInstance, Type) of
+                        false -> proper_gen:clean_instance(RawImmInstance);
+                        true -> RawImmInstance
+                      end;
+                    %% Either {ok, false} or error, suggests no user nf.
```
Comment (kostis) [inline src/proper.erl:1750]: I think it is better that this code is written in one of the following two ways:
```erlang
    Other when Other =:= {ok, false} orelse Other =:= erorr -> RawImmInstance
```
or
```erlang
   {ok, false} -> RawImmInstance;
   error -> RawImmInstance
```
because then it is robust w.r.t. what `proper_types:find_prop` returns.

File: src/proper_sa.erl [hunk 6]
```diff
@@ -76,33 +77,36 @@
...
--record(sa_target,
-        {first             = null :: proper_types:type(),
-         next              = null :: fun((_, _) -> proper_types:type()),
-         current_generated = null :: proper_gen:instance(),
-         last_generated    = null :: proper_gen:instance()}).
+-record(
+   sa_target,
+   {first             = null :: proper_types:type(),
+    next              = null :: fun((_, _) -> proper_types:type()
+                                                | proper_gen:instance()),
```
Comment (kostis) [inline src/proper_sa.erl:84]: I think it's better to declare a type for `proper_types:type() | proper_gen:instance()` and use it here.

This will also allow to undo the change in the formatting of the two record, which adds, unnecessarily, many lines of differences in this PR.

File: src/proper_target.erl [hunk 7]
```diff
@@ -258,16 +281,17 @@ init([{Strategy, Data}]) ->
 %% @private
 -spec handle_call(Request :: term(), From :: {pid(), Tag :: term()},
                   State :: state()) ->
-  {reply, Reply :: term(), NewState :: state()}.
-handle_call(gen, _From, State) ->
-  Strat = State#state.strategy,
-  Target = State#state.target,
-  Data = State#state.data,
+        {reply, Reply :: term(), NewState :: state()}.
+handle_call(gen, _From,
```
Comment (kostis) [inline src/proper_target.erl:285]: I think this looks better as:
```erlang
handle_call(gen, _From, State) ->
  #state{strategy = Strat, target = Target, data = Data} = State,
  ...
```
because it's immediately clear that there no pattern matching on the State variable.
